lib/query/query.go

```diff
@@ -0,0 +1,219 @@
+// Package query implements querying through the Salesforce data service.
+// It is a non-trivial problem because:
+//
+// - The API response shape is weird
+// - Subqueries can themselves be paginated.
+//
+// The first is generally manageable and nothing so out of the ordinary.
+// The second however is a big problem from an API/interface perspective.
+// It doesn't make sense to paginate subqueries to a caller,
+// because they probably don't have the context to use the subquery.
+// It's also just incredibly difficult to program the caller-based pagination of subqueries.
+//
+// So we choose to hide the subquery pagination as an implementation detail.
+// Again, not ideal, but probably the best choice.
+//
+// We do, however, still support top-level query pagination.
+package query
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+type result struct {
+	Done           bool
+	TotalSize      int
+	NextRecordsUrl string
+	Records        []map[string]interface{}
+}
+
+func recordsFromMapRecords(q queryer, records []map[string]interface{}) ([]Record, error) {
+	res := make([]Record, len(records))
+	for i, rec := range records {
+		rec, err := recordFromMapRecord(q, rec)
+		if err != nil {
+			return nil, err
+		}
+		res[i] = rec
+	}
+	return res, nil
+}
+
+func recordFromMapRecord(q queryer, rec map[string]interface{}) (Record, error) {
+	res := Record{
+		Raw:    rec,
+		Fields: make(map[string]interface{}, len(rec)-1),
+	}
+	for k, v := range rec {
+		if k == "attributes" {
+			attrs := v.(map[string]interface{})
+			res.Attributes.Type = attrs["type"].(string)
+			res.Attributes.Url = attrs["url"].(string)
+		} else if strings.HasSuffix(k, "__r") && v != nil {
```

Comment:
We need to check for the `records` attribute to handle relationships on standard objects, e.g. `SELECT Id, (SELECT Id FROM Contacts) FROM Account`.

Reply:
Pushed a fix.